controllers/dataprotection/backuprepo_controller.go

```diff
@@ -426,122 +516,438 @@ func (r *BackupRepoReconciler) createStorageClass(
-		paramMd5InSecret := secret.Annotations[dataProtectionTemplateValuesMD5AnnotationKey]
-		if tmplMd5InSecret == currentTemplateMd5 && paramMd5InSecret == currentParametersMd5 {
+		oldDigest := secret.Annotations[dataProtectionBackupRepoDigestAnnotationKey]
+		if oldDigest == reconCtx.getDigest() {
 			continue
 		}
 		patch := client.MergeFrom(secret.DeepCopy())
 		constructToolConfigSecret(secret, content)
 		if secret.Annotations == nil {
 			secret.Annotations = make(map[string]string)
 		}
-		secret.Annotations[dataProtectionToolConfigTemplateMD5MD5AnnotationKey] = currentTemplateMd5
-		secret.Annotations[dataProtectionTemplateValuesMD5AnnotationKey] = currentParametersMd5
-		if err = r.Client.Patch(reqCtx.Ctx, secret, patch); err != nil {
+		secret.Annotations[dataProtectionBackupRepoDigestAnnotationKey] = reconCtx.getDigest()
+		if err = r.Client.Patch(reconCtx.Ctx, secret, patch); err != nil {
 			return err
 		}
 	}
 
-	if err = updateAnnotations(reqCtx.Ctx, r.Client, repo, map[string]string{
-		dataProtectionToolConfigTemplateMD5MD5AnnotationKey: currentTemplateMd5,
-		dataProtectionTemplateValuesMD5AnnotationKey:        currentParametersMd5,
-	}); err != nil {
+	return updateAnnotations(reconCtx.Ctx, r.Client, reconCtx.repo, map[string]string{
+		dataProtectionNeedUpdateToolConfigAnnotationKey: "false",
+	})
+}
+
+func (r *BackupRepoReconciler) preCheckRepo(reconCtx *reconcileContext) (err error) {
+	if reconCtx.digestChanged() {
+		// invalidate the old status. reconCtx.preCheckFinished() depends on this value
+		err := updateCondition(reconCtx.Ctx, r.Client, reconCtx.repo, ConditionTypePreCheckPassed,
+			metav1.ConditionUnknown, ReasonDigestChanged, "")
+		if err != nil {
+			return err
+		}
+
+		err = updateAnnotations(reconCtx.Ctx, r.Client, reconCtx.repo, map[string]string{
+			dataProtectionBackupRepoDigestAnnotationKey:     reconCtx.getDigest(),
+			dataProtectionNeedUpdateToolConfigAnnotationKey: trueVal,
+		})
+		if err != nil {
+			return err
+		}
+	}
+	if reconCtx.preCheckFinished() {
+		return nil
+	}
+
+	status := metav1.ConditionUnknown
+	reason := ReasonUnknownError
+	message := ""
+	defer func() {
+		if message == "" && err != nil {
+			message = err.Error()
+		}
+		r.updateConditionInDefer(reconCtx.Ctx, reconCtx.repo, ConditionTypePreCheckPassed, reason, &status, &message, &err)
+	}()
+	var job *batchv1.Job
+	var pvc *corev1.PersistentVolumeClaim
+	switch {
+	case reconCtx.repo.AccessByMount():
+		job, pvc, err = r.runPreCheckJobForMounting(reconCtx)
+	case reconCtx.repo.AccessByTool():
+		job, err = r.runPreCheckJobForTool(reconCtx)
+	default:
+		err = fmt.Errorf("unknown access method: %s", reconCtx.repo.Spec.AccessMethod)
+	}
+	if err != nil {
+		return err
+	}
+
+	finished, jobStatus, failureReason := utils.IsJobFinished(job)
+	if !finished {
+		duration := wallClock.Since(job.CreationTimestamp.Time)
+		if duration > defaultPreCheckTimeout {
```

Comment:
Can we use the job's `activeDeadlineSeconds` filed? [activeDeadlineSeconds](https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-termination-and-cleanup)

Reply:
Job and Pod both have activeDeadlineSeconds, but neither of them is suitable for our scenario. If `job.spec.activeDeadlineSeconds` is set, when the run times out, the job controller will delete the running pods directly to stop them; since the pods are deleted, we may not have time to collect the error logs.

In the meantime, `pod.spec.activeDeadlineSeconds` may fail in some cases. As mentioned before, when the configuration is wrong, the PVC provisioning will fail, which makes the pod get stuck in the `Pending` state, but activeDeadlineSeconds seems to start counting from the `Running` state, so the pod will not fail due to timeout.

Reply:
Ok, It is recommended to add this description to the code comments.

Reply:
Done.